propagators/autoprop/registry.go

```diff
@@ -0,0 +1,90 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autoprop // import "go.opentelemetry.io/contrib/propagators/autoprop"
+
+import (
+	"sync"
+
+	"go.opentelemetry.io/contrib/propagators/aws/xray"
+	"go.opentelemetry.io/contrib/propagators/b3"
+	"go.opentelemetry.io/contrib/propagators/jaeger"
+	"go.opentelemetry.io/contrib/propagators/ot"
+	"go.opentelemetry.io/otel/propagation"
+)
+
+// none is the special "propagator" name that means no propagator shall be
+// configured.
+const none = "none"
+
+// envRegistry is the index of all supported environment variable
+// values and their mapping to a TextMapPropagator.
+var envRegistry = &registry{
+	names: map[string]propagation.TextMapPropagator{
+		// W3C Trace Context.
+		"tracecontext": propagation.TraceContext{},
+		// W3C Baggage.
+		"baggage": propagation.Baggage{},
+		// B3 single-header format.
+		"b3": b3.New(),
+		// B3 multi-header format.
+		"b3multi": b3.New(b3.WithInjectEncoding(b3.B3MultipleHeader)),
+		// Jaeger.
+		"jaeger": jaeger.Jaeger{},
+		// AWS X-Ray.
+		"xray": xray.Propagator{},
+		// OpenTracing Trace.
+		"ottrace": ot.OT{},
+
+		// No-op TextMapPropagator.
+		none: propagation.NewCompositeTextMapPropagator(),
+	},
+}
+
+// registry maintains a map of propagator names to TextMapPropagator
+// implementations that is safe for concurrent use by multiple goroutines
+// without additional locking or coordination.
+type registry struct {
+	mu    sync.Mutex
+	names map[string]propagation.TextMapPropagator
+}
+
+// load returns the value stored in the registry index for a key, or nil if no
+// value is present. The ok result indicates whether value was found in the
+// index.
+func (r *registry) load(key string) (p propagation.TextMapPropagator, ok bool) {
+	r.mu.Lock()
+	p, ok = r.names[key]
+	r.mu.Unlock()
+	return p, ok
+}
+
+// store sets the value for a key.
+func (r *registry) store(key string, value propagation.TextMapPropagator) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.names == nil {
+		r.names = map[string]propagation.TextMapPropagator{key: value}
+		return
+	}
+	r.names[key] = value
+}
+
+// RegisterTextMapPropagator sets the TextMapPropagator p to be used when the
+// OTEL_PROPAGATORS environment variable contains the propagator name. This
+// allows the default supported environment TextMapPropagators to be extended
+// with 3rd-part implementations.
+func RegisterTextMapPropagator(name string, p propagation.TextMapPropagator) {
+	envRegistry.store(name, p)
```

Comment:
Do we care that this could override existing propagators, or that there can be naming conflicts?

Reply:
It was intentional to allow overrides of defaults or existing registrations. This would allow 3rd party implementations to be used instead of the default propagators. I tried to capture this intention in the function docs, does it need to be made clearer?

As for the name collision, I was assuming if a user registered a propagator and used that name they would get what they expected: the name collision is resolved by the override being returned. This seemed like the intuitive behavior, but I'm open to suggestions.

Reply:
Ah, I read "extended with 3rd-part implementations" as meaning you can add additional third party propagators.  Maybe "This allows adding additional environment TextMapPropagators, or replacing the default TextMapPropagators with 3rd-party implementations."

(also s/part/party)

Reply:
Docs updated.

Reply:
I'm wary of allowing overriding propagator implementations that already have well-known names.  That seems like a recipe for user confusion.  I think that if someone wants to register a third-party implementation they should register it with a new name.

Reply:
I imagine people (or even distributions) may want to provide their own implementation of given propagator e.g. for bugfixing, better logging, improving performance etc

Reply:
> I'm wary of allowing overriding propagator implementations that already have well-known names. That seems like a recipe for user confusion.

My thought here was that it would be the user doing the registration and therefore the user would know there was an override.

Though, this comment made me wonder if there could be a situation where a library that a user is using in addition to `autoprop` that sets and override without the users knowledge. I imagine that library would document their registration if they were a good actor, but if they were a bad actor it might be done maliciously.

The alternative, where a user will use a different name for a similar propagator might be a better solution with this in mind. It makes the alternate implementation use explicit and the configuration of propagators one-sided (only with environment variables).

What do you think the behavior should be on a duplicate registration? Just drop?

Reply:
From SIG meeting:

- Dropping the attempt to double register still leaves open the possibility of a malicious user registering their `init` prior to a users' code and having the users' code silently not register. This is not the behavior wanted.
- Panicking seems the safest and will clearly tell the user when there are conflicts.
  - The downside is that non-malicious just accidental collisions cannot be ignored. For instance, two packages register the same propagator with the same name and the user imports them both. The user does not care about the conflict, they just care that one was successful.
- Return an error and allow the user to handle the error.
   - Allowing the user to decide how to handle conflicts allows the most flexibility and still provides _some_ security.
   - Errors are often ignored or unhanded (just logged). This might not be the right user interface given the sensitive data a propagator could have access to.
   - Even if an error is returned, it is going to be the dependent library that needs to appropriately handle this, not the user importing that library. The only safe way for that library to handle the situation would seem for them to panic themselves. Which would indicate that returning an error here only pushes the security concern to dependencies (who may not recognize it as such and not handle it appropriately).

I'm planning to change to a panic based on this. Open to alternative opinions though.

Reply:
Updated, please take another look.